File: message-bus/src/test/java/alma/obops/draws/messages/security/MockedTokenFactory.java
package alma.obops.draws.messages.security;

import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

public class MockedTokenFactory implements TokenFactory {

	private static final String HEADER = "header";
	private static final String VALID_SIGNATURE = "valid-signature";
	private static final String INVALID_SIGNATURE = "invalid-signature";

	protected static MockedTokenFactory instance = null;
	
	public static TokenFactory getFactory() {
		if( instance == null ) {
			instance = new MockedTokenFactory(); 
		}
		return instance;
	}
	
	private MockedTokenFactory() {
	}

	@Override
	public String create() {
		Map<String, Object> claims = new HashMap<>();
		claims.put( "sub", "user" );
		claims.put( "role", "admin" );
		claims.put( "ttl", "10000" );
		return create( claims );
	}

	@Override
	public String create( Map<String, Object> properties ) {
		if( properties == null ) {
			throw new IllegalArgumentException( "No properties given" );
		}
		String header = HEADER;
		
		String signature = VALID_SIGNATURE;
		Object valid = properties.get( "valid" );
		if( valid != null && valid.equals( "false" )) {
			signature = INVALID_SIGNATURE;
		}
		StringBuilder bodyBuilder = new StringBuilder();
		for( String propName : properties.keySet() ) {
			Object propValue = properties.get( propName );
			if( bodyBuilder.length() > 0 ) {
				bodyBuilder.append( "," );
			}
			bodyBuilder.append( propName ).append( "=" ).append( propValue );
		}
		
		String body = bodyBuilder.toString();
		String encodedBody = Base64.getEncoder().encodeToString( body.getBytes() );
		return header + "." + encodedBody + "." + signature;
	}

	@Override
	public Map<String,Object> decode( String token ) throws InvalidSignatureException {

		final String encodedBody = isValidInternal( token );
		if( encodedBody == null ) {
			throw new InvalidSignatureException();
		}
		
		String body = new String( Base64.getDecoder().decode( encodedBody ));
		Map<String, Object> properties = new HashMap<>();
		String[] claims = body.split( "," );
		for( int i = 0; i < claims.length; i++ ) {
			String  claim = claims[i];
			String[]  t2 = claim.split( "=" );
			String   key = t2[0];
			String value = t2[1];
			
			properties.put(key, value);
		}
		
		return properties;
	}

	@Override
	public boolean isValid( String token ) {
		return isValidInternal( token ) != null;
	}

	/**
	 * @return <code>null</code> if the input token is invalid, the token body
	 *         otherwise
	 */
	private String isValidInternal( String token ) {
		if( token == null || token.length() == 0 ) {
			return null;
		}
		
		String[] t = token.split( "\\." );
		if( t.length != 3 ) {
			return null;
		}

//		String header    = t[0];	IGNORED
		String body      = t[1];
		String signature = t[2];
		
		if( ! signature.equals( VALID_SIGNATURE )) {
			return null;
		}
		return body;
	}
}

File: message-bus/src/main/python/draws/test/messages/security/MockedTokenFactory.py
from draws.messages.security.TokenFactory import TokenFactory

import base64

class MockedTokenFactory(TokenFactory):
    #Static variables
    __HEADER = "header"
    __VALID_SIGNATURE = "valid-signature"
    __INVALID_SIGNATURE = "invalid-signature"
    _instance = None

    #Static Methods
    @classmethod
    def getFactory(cls):
        if MockedTokenFactory._instance is None:
            MockedTokenFactory._instance = MockedTokenFactory()
        return MockedTokenFactory._instance

    #Instance methods
    def create(self, properties=None):
        if properties is None:
            claims = {}
            claims["sub"] = "user"
            claims["role"] = "admin"
            claims["ttl"] = "10000"
            return self.create(claims)
        if properties is None:
            raise IllegalArgumentException( "No properties given" )
        header = MockedTokenFactory.__HEADER
        signature = MockedTokenFactory.__VALID_SIGNATURE
        valid = None
        if "valid" in properties:
            valid = properties["valid"]
        if valid is not None and valid == "false":
            signature = MockedTokenFactory.__INVALID_SIGNATURE
        bodyBuilder = ""
        for propName in properties:
            propValue = properties.get(propName)
            if len(bodyBuilder) > 0:
                bodyBuilder += ","
            bodyBuilder += propName + "=" + propValue
        body = bodyBuilder
        encodedBody = base64.b64encode(body.encode("utf-8")).decode()
        return header + "." + encodedBody + "." + signature

    def decode(self, token):
        encodedBody = self.__isValidInternal( token )
        if encodedBody is None:
            raise Exception("Invalid signature...")
        body = base64.b64decode(encodedBody.encode()).decode()
        properties = {}
        claims = body.split(",")
        for claim in claims:
            t2 = claim.split( "=" )
            properties[t2[0]] = t2[1]
        return properties

    def isValid(self, token):
        return self.__isValidInternal(token) is not None

    def __isValidInternal(self, token):
        if token is None or len(token) == 0:
            return None
        
        t = token.split(".")
        if len(t) != 3:
            return None

#       header    = t[0]    IGNORED
        body      = t[1]
        signature = t[2]
        
        if signature != MockedTokenFactory.__VALID_SIGNATURE:
            return None
        return body
